rllib/core/testing/torch/marl_module_obs_encoder.py

```diff
@@ -0,0 +1,138 @@
+        self, batch: MultiAgentBatch, module_id: ModuleID = "", **kwargs
+    ) -> Union[Mapping[str, Any], Dict[ModuleID, Mapping[str, Any]]]:
+        """Runs the forward_train pass.
+
+        Args:
+            batch: The batch of multi-agent data (i.e. mapping from module ids to
+                SampleBaches).
+            module_id: The module ID to run the forward pass for. If not specified, all
+                modules are run.
+
+        Returns:
+            The output of the forward_train pass the specified modules.
+        """
+        return self.__run_forward_pass("forward_train", batch, module_id, **kwargs)
+
+    @override(RLModule)
+    def _forward_inference(
+        self, batch: MultiAgentBatch, module_id: ModuleID = "", **kwargs
+    ) -> Union[Mapping[str, Any], Dict[ModuleID, Mapping[str, Any]]]:
+        """Runs the forward_inference pass.
+
+        Args:
+            batch: The batch of multi-agent data (i.e. mapping from module ids to
+                SampleBaches).
+            module_id: The module ID to run the forward pass for. If not specified, all
+                modules are run.
+
+        Returns:
+            The output of the forward_inference pass the specified modules.
+        """
+        return self.__run_forward_pass("forward_inference", batch, module_id, **kwargs)
+
+    @override(RLModule)
+    def _forward_exploration(
+        self, batch: MultiAgentBatch, module_id: ModuleID = "", **kwargs
+    ) -> Union[Mapping[str, Any], Dict[ModuleID, Mapping[str, Any]]]:
+        """Runs the forward_exploration pass.
+
+        Args:
+            batch: The batch of multi-agent data (i.e. mapping from module ids to
+                SampleBaches).
+            module_id: The module ID to run the forward pass for. If not specified, all
+                modules are run.
+
+        Returns:
+            The output of the forward_exploration pass the specified modules.
+        """
+        return self.__run_forward_pass(
+            "forward_exploration", batch, module_id, **kwargs
+        )
+
+    def __run_forward_pass(
+        self,
+        forward_fn_name: str,
+        batch: NestedDict,
+        module_id: ModuleID = "",
+        **kwargs,
+    ) -> Dict[ModuleID, Mapping[str, Any]]:
+        for _module_id in batch.shallow_keys():
+            agent_batch = batch[_module_id]
+            agent_batch["obs"] = self.encoder(agent_batch["obs"])
+
+        module_ids = self.keys() if module_id == "" else [module_id]
+
+        outputs = {}
+        for module_id in module_ids:
+            rl_module = self._rl_modules[module_id]
+            forward_fn = getattr(rl_module, forward_fn_name)
+            outputs[module_id] = forward_fn(batch[module_id], **kwargs)
+
+        return outputs
+
+    @override(MultiAgentRLModule)
+    def input_specs_exploration(self) -> SpecDict:
+        sd = super().input_specs_exploration()
+        for module_id in sd.keys():
+            if "obs" in module_id:
```

Comment:
I did this to avoid having to write my own Nested tensor specs. In hindsight this looks p complicated.